File: cpp/189_RotateArray.cpp
class Solution {
public:
    void rotate(vector<int>& nums, int k) {
        k = k % nums.size();
        if (k == 0) return;
        
        int count = 0;
        for (int start = 0; count < nums.size(); ++start) {
            int curr = start;
            int prev = nums[start];
            do {
                int next = (curr + k) % nums.size();
                int temp = nums[next];
                nums[next] = prev, prev = temp, curr = next;
                count++;
            } while (start != curr);
        }
    }
};

File: swift/189_RotateArray.swift
class Solution {
    func rotate(_ nums: inout [Int], _ k: Int) {
        let N = nums.count, jump = k % N
        if jump == 0 { return }
        
        var start = 0, count = 0
        while count < N {
            var currentIteration = start, previousNum = nums[start]
            repeat {
                var next = (start + jump) % N, temp = nums[next]
                nums[next] = previousNum
                previousNum = temp
                start = next
                count += 1
            } while start != currentIteration
            start += 1
        }
    }
}
